plugins/net_plugin/net_plugin.cpp

```diff
@@ -3136,39 +3134,60 @@ namespace eosio {
    void connection::handle_message( const block_id_type& id, signed_block_ptr ptr ) {
       peer_dlog( this, "received signed_block ${num}, id ${id}", ("num", ptr->block_num())("id", id) );
 
-      controller& cc = my_impl->chain_plug->chain();
-      block_state_ptr bsp;
-      bool exception = false;
-      try {
-         if( cc.fetch_block_state_by_id( id ) ) {
-            my_impl->dispatcher->add_peer_block( id, connection_id );
-            my_impl->sync_master->sync_recv_block( shared_from_this(), id, ptr->block_num(), false );
+      // post to dispatcher strand so that we don't have multiple threads validating the block header
+      // the dispatcher strand will sync the add_peer_block and rm_block calls
+      my_impl->dispatcher->strand.post([id, c{shared_from_this()}, ptr{std::move(ptr)}, cid=connection_id]() mutable {
+         controller& cc = my_impl->chain_plug->chain();
+
+         // may have come in on a different connection and posted into dispatcher strand before this one
+         if( my_impl->dispatcher->have_block( id ) || cc.fetch_block_state_by_id( id ) ) {
+            my_impl->dispatcher->add_peer_block( id, c->connection_id );
+            c->strand.post( [c, id, blk_num=ptr->block_num()]() {
+               my_impl->sync_master->sync_recv_block( c, id, blk_num, false );
+            });
             return;
          }
-         // this may return null if block is not immediately ready to be processed
-         bsp = cc.create_block_state( id, ptr );
-      } catch( const fc::exception& ex ) {
-         exception = true;
-         peer_elog(this, "bad block exception: #${n} ${id}...: ${m}",
-                   ("n", ptr->block_num())("id", id.str().substr(8,16))("m",ex.to_string()));
-      } catch( ... ) {
-         exception = true;
-         peer_elog(this, "bad block: #${n} ${id}...: unknown exception",
-                   ("n", ptr->block_num())("id", id.str().substr(8,16)));
-      }
-      if( exception ) {
-         my_impl->sync_master->rejected_block( shared_from_this(), ptr->block_num() );
-         my_impl->dispatcher->rejected_block( id );
-         return;
-      }
 
-      bool signal_producer = !!bsp; // ready to process immediately, so signal producer to interrupt start_block
-      app().post(priority::medium, [ptr{std::move(ptr)}, bsp{std::move(bsp)}, id, c = shared_from_this()]() mutable {
-         c->process_signed_block( id, std::move(ptr), std::move(bsp) );
-      });
+         block_state_ptr bsp;
+         bool exception = false;
```

Comment:
Why not just check `bsp` and forego `bool exception;`?

Reply:
Because an exception happening is different than `create_block_state` returning an empty `block_state_ptr`. An empty `block_state_ptr` means the block is not ready to be processed right now, but might be soon. An exception indicates there is something wrong with this block. An exception we know we can immediately reject the block. An empty `block_state_ptr` means we need to post it for later processing.

Reply:
Got it, I should read the comments more carefully!